src/main/groovy/wooga/gradle/paket/unity/tasks/PaketUnwrapUPMPackages.groovy

```diff
@@ -0,0 +1,199 @@
+ * <p>
+ * Example:
+ * <pre>
+ * {@code
+ *     task unwrapUpmPackages(type:wooga.gradle.paket.unity.tasks.PaketUnwrapUPMPackages) {
+ *         referencesFile = file('paket.unity3D.references')
+ *         lockFile = file('../paket.lock')
+ *     }
+ * }
+ * </pre>
+ */
+class PaketUnwrapUPMPackages extends ConventionTask {
+
+    /**
+     * @return the path to a {@code paket.unity3d.references} file
+     */
+    @InputFile
+    File referencesFile
+
+    /**
+     * @return the path to a {@code paket.lock} file
+     */
+    @InputFile
+    File lockFile
+
+    public final static String localUPMWrapperPackagePrefix = "Wooga.UPMWrapper"
+
+    /**
+     * @return the installation output directory
+     */
+    @OutputDirectory
+    File getOutputDirectory() {
+        new File(getReferencesFile().getParentFile(), "Packages")
+    }
+
+    /**
+     * @return the paket.upm.wrapper.reference's of the wrapper packages to unwrap into the Unity3D/Packages project.
+     */
+    @InputFiles
+    FileCollection getInputFiles() {
+        Set<File> files = []
+        def references = new PaketUnityReferences(getReferencesFile())
+
+        if (!getLockFile().exists()) {
+            return null
+        }
+
+        def locks = new PaketLock(getLockFile())
+        def dependencies = locks.getAllDependencies(references.nugets)
+        dependencies.each { nuget ->
+            def depFiles = getFilesForPackage(nuget)
+            files << depFiles
+        }
+
+        project.files(files)
+    }
+
+    Set<File> getFilesForPackage(String nuget) {
+        if (isUPMWrapper(nuget)) {
+            return [new PaketUPMWrapperReference(nuget, project).file] as Set<File>
+        }
+    }
+
+    PaketUnwrapUPMPackages() {
+        description = 'Unwraps wrapped UPM into paket dependencies into unity projects as local UPM Overrides'
+        group = PaketUnityPlugin.GROUP
+    }
+
+    @TaskAction
+    protected performCopy(IncrementalTaskInputs inputs) {
+        if (!inputs.incremental) {
+            cleanLocalUpmOverrides()
+        }
+
+        inputs.outOfDate(new Action<InputFileDetails>() {
+            @Override
+            void execute(InputFileDetails outOfDate) {
+                if(inputFiles.contains(outOfDate.file) && PaketUPMWrapperReference.IsReferenceFile(outOfDate.file)) {
+                    def upmWrapperRef = new PaketUPMWrapperReference(outOfDate.file)
+                    if (upmWrapperRef.exists) {
+                        cleanLocalUpmOverride(upmWrapperRef)
+                        def upmUnwrapper = upmWrapperRef.upmPackageURL.startsWith("http")
+                                ? new DownloadAndUnpackTar(upmWrapperRef, getOutputDirectory())
+                                : new UnwrapUpm(upmWrapperRef, getOutputDirectory())
+                        upmUnwrapper.exec(project)
```

Comment:
So it support remote and package local files?

Reply:
well i stared with remote, then the unity tgz packaging bug made me pivot, but didn't want to throw away this possibility since it was already sitting there :)